devtools/src/arcs-hot-reload.js

```diff
@@ -0,0 +1,149 @@
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {PolymerElement} from '../deps/@polymer/polymer/polymer-element.js';
+import {html} from '../deps/@polymer/polymer/lib/utils/html-tag.js';
+import {MessengerMixin} from './arcs-shared.js';
+
+class HotReload extends MessengerMixin(PolymerElement) {
+  static get template() {
+    return html`
+    <style include="shared-styles">
+      :host {
+        display: flex;
+        flex-direction: column;
+      }
+      .switch {
+        position: relative;
+        display: inline-block;
+        width: 30px;
+        height: 17px;
+      }
+      .slider {
+        position: absolute;
+        top: 0;
+        left: 0;
+        right: 0;
+        bottom: 0;
+        background-color: #ccc;
+        border-radius: 34px;
+      }
+      .slider:before {
+        position: absolute;
+        content: "";
+        height: 13px;
+        width: 13px;
+        left: 2px;
+        bottom: 2px;
+        background-color: white;
+        -webkit-transition: .4s;
+        border-radius: 50%;
+      }
+      input:checked + .slider {
+        background-color: #AFF196;
+      }
+      input:checked + .slider:before {
+        -webkit-transform: translateX(13px);
+      }
+      .tooltip:hover .tooltiptext {
+        visibility: visible;
+      }
+      .tooltip .tooltiptext {
+        visibility: hidden;
+        width: 120px;
+        text-align: center;
+        color: #838783;
+        font-size: 10px;
+        position: absolute;
+        top: 8px;
+        left: 45%;
+      }
+      .tooltip {
+        position: relative;
+        display: inline-block;
+      }
+    </style>
+    <div class="tooltip">
+      <label class="switch">
+        <input type="checkbox" on-click="_loadServer" id="toggle">
+        <span class="slider round"></span>
+      </label>
+      <span class="tooltiptext">
+        Hot Code Reload
+      </span>
+    </div>`;
+  }
+
+  static get is() { return 'arcs-hot-reload'; }
+
+  constructor() {
+    super();
+    this.hotReloadWS = null;
+    this.hotReloadServer = null;
+    this.particleSources = new Set();
+    this.arcIdsToKeepFresh = [];
+  }
+
+  onMessage(msg) {
+    switch (msg.messageType) {
+      case 'register-arc':
+        this.arcIdsToKeepFresh.push(msg.messageBody);
+        break;
+      case 'register-particle-sources': {
+        const before = this.particleSources.size;
+        for (const particleSource of msg.messageBody) {
+          this.particleSources.add(particleSource);
+        }
+        if (this.particleSources.size > before && this.hotReloadWS != null) {
+          const files = JSON.stringify([...this.particleSources]);
+          this.hotReloadWS.send(files);
+        } 
+        break;
+      }
+      case 'particle-reload-result':
+        break;
+    }
+  }
+
+  _loadServer() {
+    if (this.$.toggle.checked) {
+      this.$.toggle.checked = false;
+
+      this.hotReloadWS = new WebSocket('ws://localhost:8888');
```

Comment:
I don't understand. We create a WebSocket only when the toggle was checked when we clicked it?

I would expect it to be the other way around: the toggle is unchecked, the user clicks it, we mark it as checked and open a WebSocket

Reply:
This is because whenever we click the toggle then its 'checked' field will automatically changed to true. I changed it back to false since we want to check whether we can actually connect to ALDS first before changing the toggle display. (in case there is no ALDS and the connection failed then our toggle display won't change).

Reply:
Oooo! Tricky business! :)